_posts/2017-09-20-presentation-protocol-buffer.md

```diff
@@ -0,0 +1,254 @@
+authors: 
+    - qneyrat
+permalink: /fr/presentation-protocol-buffer/
+categories:
+    - Api
+    - Go
+tags:
+    - api
+    - go
+---
+
+Une problématique assez récurrente dans nos applications est la manière de rendre notre donnée, qu'elle soit transmise ou stockée. Le format qu'on utilise souvent aujourd'hui est le `JSON`. Cependant certain language le gère très mal tel que Java et Go. Google a donc développé un système pour palier à ce problème : `Protocol Buffer`.
+
+## Présentation
+---
+
+Protocol Buffer est un système de sérialisation de données tout comme `JSON` et `XML`. Il est disponible pour la plupart des languages. Une application en Java peut envoyer des objets à une application en Go. Le système repose sur un fichier qui va permettre de structurer notre objet, les fichiers `.proto`. Ce fichier va un peu comme une interface décrire notre objet. Protobuf permet ensuite de générer le code source de l'objet dans plusieurs langage différent.
+
+Pour récapituler, on déclare un fichier proto, on génère notre objet dans notre application serveur et dans notre application client. Nos objets auront dans leurs déclaration des méthodes de sérialisation et de de-sérialisation et ce quelque soit le langage.
+
+Notre exemple va être le suivant :
+
+Notre api va retourné un objet `Post`. Un client va appeler cette api. Nous allons avoir besoin d'un fichier `proto` qui va générer le code source en Go. Le serveur va sérialisé un objet et le rendre au client. Ce qui nous donne :
+
+```
+Go Struct  ↘                                              ↗ Go Struct
+             Serialization -> bytes -> Deserialization 
+Proto file ↗                                              ↖ Proto file     
+```
+
+Nous allons maintenant voir étape par étape comment ça fonctionne.
+
+## Fonctionnement
+---
+
+Protobuf est un language qui va permettre de définir comment l'objet va être sérialisé et comment il va générer le code source.
+Voici un example de fichier protobuf :
+
+```proto
+message Person {
+  string name = 1;
+  int32 id = 2;
+  string email = 3;
+
+  enum PhoneType {
+    MOBILE = 0;
+    HOME = 1;
+    WORK = 2;
+  }
+
+  message PhoneNumber {
+    string number = 1;
+    PhoneType type = 2;
+  }
+
+  repeated PhoneNumber phones = 4;
+}
+```
+
+Dans cette example notre objet `Person` est constitué d'un `name`, d'un `id`, d'un `email` et de `phones`.
+La déclaration d'une propriété est définie par un type `int32` ou `string` (et bien d'autres), du nom de la propriété puis d'un identifiant unique (la position) de cette propriété.
+
+On peut aussi faire de la composition est créant de nouveaux types comme ici `PhoneNumber` ou bien des énumérations.
+Des modèles de données sont aussi disponible comme les `array` avec le mot clé `repeated` ou bien encore des maps avec `map<Key, Value>`.
+
+Vous pouvez retrouver tout les types et déclaration sur [la documentation de Protobuf](https://developers.google.com/protocol-buffers/docs/proto3).
```

Comment:
`tout les types` -> `tous les types`

Reply:
You can add this as a suffix if you want your link to be open in a new tab: `{:target="_blank"}`